### ISSUE TITLE
Use fp16 performance decrease

### ISSUE THREAD

**Issue by mx8435** [NVIDIA/apex#595]
Hi, I use apex op_level O2, and do some cast in some places, but the performance decrease.

![image](https://user-images.githubusercontent.com/8417466/68458594-24fcda00-023e-11ea-9425-6b086cdb2d13.png)

Any way to find where the precision decrease by use fp16 (or cast float to float16)?

Besides, will `FFNLayer(3 * hidden_size, hidden_size, 2, dropout_prob)`only accept float16 input when using `amp`?

**Comment by ptrblck** [comment]
@mx8435 Is this behavior reproducible and did you make sure e.g. set the seed as described [here](https://pytorch.org/docs/stable/notes/randomness.html)?
Could you rerun the training with and without amp and check the stddev of the final accuracy?

**Comment by mx8435** [comment]
> @mx8435 Is this behavior reproducible and did you make sure e.g. set the seed as described [here](https://pytorch.org/docs/stable/notes/randomness.html)?
> Could you rerun the training with and without amp and check the stddev of the final accuracy?

Thanks for your reply. 
The code has set `torch.manual_seed` and `np.random.seed` but not CuDNN options.
My final result using amp is 81.16 while the final result is 83.76 without amp, I think is a big decrease. I reran the code with `amp` twice, and got 81.23.

**Comment by BramVanroy** [comment]
Only using numpy and torch seeds is not sufficient. This might help:

```python
    def _set_seed(seed):
        """ Set all seeds to make results reproducible (deterministic mode).
            When seed is None, disables deterministic mode. """

        if seed is not None:
            torch.manual_seed(seed)
            torch.cuda.manual_seed_all(seed)
            torch.backends.cudnn.deterministic = True
            torch.backends.cudnn.benchmark = False
            np.random.seed(seed)
            random.seed(seed)
            os.environ['PYTHONHASHSEED'] = str(seed)
```

**Comment by mx8435** [comment]
> stddev

I've reran the code without amp, I think the variance is minor.
![image](https://user-images.githubusercontent.com/8417466/68721274-d27c3e80-05ec-11ea-804e-c2fdf2af32d8.png)

Any way to find where the precision decrease caused by fp16 (or by call `tensor.half()`)?
